examples/control-vector-generator/control-vector-generator.cpp

```diff
@@ -0,0 +1,839 @@
+    struct ggml_tensor * t_host;
+    auto n_bytes = ggml_nbytes(t);
+    t_host = ggml_new_tensor_2d(cb_data->ctx_ggml, t->type, t->ne[0], t->ne[1]);
+    t_host->data = malloc(n_bytes); // TODO @ngxson : get rid of this malloc somehow
+    ggml_backend_tensor_get(t, t_host->data, 0, n_bytes);
+    printf("t_host [0][%d]: %f\n", DEBUG_POS, ggml_get_f32_nd(t_host, 0, DEBUG_POS, 0, 0));
+
+    if (t_host->type == GGML_TYPE_F32) {
+        if (cb_data->is_eval_pos) {
+            cb_data->v_pos.push_back(t_host);
+        } else {
+            cb_data->v_neg.push_back(t_host);
+        }
+    }
+
+    return true;
+}
+
+static bool get_hidden_layers(llama_context * ctx, std::vector<llama_token> & tokens) {
+    llama_kv_cache_clear(ctx);
+    if (llama_decode(ctx, llama_batch_get_one(tokens.data(), tokens.size(), 0, 0))) {
+        fprintf(stderr, "%s : failed to eval\n", __func__);
+        return false;
+    }
+    return true;
+}
+
+static void padding_seq(llama_context * ctx, std::vector<llama_token> & tokens, size_t len) {
+    // TODO: customize padding token
+    std::vector<llama_token> pad_tokens = ::llama_tokenize(ctx, " ", false);
+    llama_token pad_tok = pad_tokens.back();
+    while (tokens.size() < len) {
+        tokens.push_back(pad_tok);
+    }
+}
+
+static bool is_row_all_zeros(struct ggml_tensor * diff, int row, int cols, float eps = 1e-6) {
+    for (int i = 0; i < cols; ++i) {
+        if (ggml_get_f32_nd(diff, i, row, 0, 0) > eps) {
+            return false;
+        }
+    }
+    return true;
+}
+
+static void calc_diff(callback_data & cb_data) {
+    // TODO: assert cb_data.v_pos.size() == cb_data.v_neg.size()
+    cb_data.v_diffs_wrapped.resize(cb_data.v_pos.size());
+    for (size_t il = 0; il < cb_data.v_pos.size(); il++) {
+        auto & inp_pos = cb_data.v_pos[il];
+        auto & inp_neg = cb_data.v_neg[il];
+        auto n_bytes = ggml_nbytes(inp_pos);
+
+        struct ggml_init_params params = {
+            /*.mem_size   =*/ n_bytes + ggml_tensor_overhead(),
+            /*.mem_buffer =*/ NULL,
+            /*.no_alloc   =*/ false,
+        };
+        struct ggml_context * ctx_data = ggml_init(params);
+
+        printf("inp_pos [0][%d]: %f\n", DEBUG_POS, ggml_get_f32_nd(inp_pos, 0, DEBUG_POS, 0, 0));
+        printf("inp_neg [0][%d]: %f\n", DEBUG_POS, ggml_get_f32_nd(inp_neg, 0, DEBUG_POS, 0, 0));
+
+        // TODO is this the best way to get dimension? i don't know which way n_embd/n_tokens go
+        // for that matter can we get rid of n_embd/n_tokens fields in favor of ne[0]/ne[1]?
+        struct ggml_tensor * dest = ggml_new_tensor_2d(ctx_data, GGML_TYPE_F32, inp_pos->ne[0], inp_pos->ne[1]);
+    
+        for (size_t i = 0; i < cb_data.n_embd; i++) {
+            for (size_t j = 0; j < cb_data.n_tokens; j++) {
+                ggml_set_f32_nd(dest, i, j, 0, 0, ggml_get_f32_nd(inp_pos, i, j, 0, 0) - ggml_get_f32_nd(inp_neg, i, j, 0, 0));
```

Comment:
Also, is there a matrix subtraction operation in the GGML library that could speed this up? Would it be worth implementing?

Reply:
Matrix add / sub are very fast even on CPU, so I don't think it's worth to implement ggml here. The time it take to just copy out the tensor to ggml can be even longer than just doing simple single-threaded sub here.

Reply:
Thought so, thanks.